RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZT.cc

```diff
@@ -0,0 +1,778 @@
+    k->swE = 0.;  k->Tc=0.;
+  }
+
+
+  // loop over tracks
+  for(unsigned int i=0; i<nt; i++){
+
+    // update pik and Zi and Ti
+    double Zi = rho0*std::exp(-beta*(dzCutOff_*dzCutOff_));// cut-off (eventually add finite size in time)
+    //double Ti = 0.; // dt0*std::exp(-beta*dtCutOff_);
+    for(vector<vertex_t>::iterator k=y.begin(); k!=y.end(); k++){
+      k->ei = std::exp(-beta*Eik(tks[i],*k));// cache exponential for one track at a time
+      Zi   += k->pk * k->ei;
+    }
+    tks[i].Z=Zi;
+    
+    // normalization
+    if (tks[i].Z>0){
+       // accumulate weighted z and weights for vertex update
+      for(vector<vertex_t>::iterator k=y.begin(); k!=y.end(); k++){
+	k->se += tks[i].pi* k->ei / Zi;
+	double w = k->pk * tks[i].pi * k->ei /( Zi * ( tks[i].dz2 * tks[i].dt2 ) );
+	k->sw  += w;
+	k->swz += w * tks[i].z;
+        k->swt += w * tks[i].t;
+	k->swE += w * Eik(tks[i],*k);
+      }
+    }
+
+
+  } // end of track loop
+
+
+  // now update z
+  double delta=0;
+  for(vector<vertex_t>::iterator k=y.begin(); k!=y.end(); k++){
+    if ( k->sw > 0){
+      const double znew=k->swz/k->sw; 
+      const double tnew=k->swt/k->sw;
+      delta += std::pow(k->z-znew,2.) + std::pow(k->t-tnew,2.);
+      k->z   = znew;
+      k->t   = tnew;
+      k->Tc  = 2*k->swE/k->sw;
+    }else{
+      edm::LogInfo("sumw") <<  "invalid sum of weights in fit: " << k->sw << endl;
+      if(verbose_){cout << " a cluster melted away ?  pk=" << k->pk <<  " sumw=" << k->sw <<  endl;}
+      k->Tc = 0;
+    }
+
+  }
+
+  // return how much the prototypes moved
+  return delta;
+}
+
+bool DAClusterizerInZT::merge(vector<vertex_t> & y, int nt)const{
+  // merge clusters that collapsed or never separated, return true if vertices were merged, false otherwise
+  
+  if(y.size()<2)  return false;
+  
+  for(vector<vertex_t>::iterator k=y.begin(); (k+1)!=y.end(); k++){
+    if( std::abs( (k+1)->z - k->z ) < 1.e-3 &&
+        std::abs( (k+1)->t - k->t ) < 1.e-3    ){  // with fabs if only called after freeze-out (splitAll() at highter T)
+      double rho = k->pk + (k+1)->pk;
+      if(rho>0){
+        k->z = ( k->pk * k->z + (k+1)->z * (k+1)->pk)/rho;
+        k->t = ( k->pk * k->t + (k+1)->t * (k+1)->pk)/rho;
+      }else{
+        k->z = 0.5*(k->z + (k+1)->z);
+        k->t = 0.5*(k->t + (k+1)->t);
+      }
+      k->pk = rho;
+      
+      y.erase(k+1);
+      return true;  
+    }
+  }
+  
+  return false;
+}
+
+bool DAClusterizerInZT::merge(vector<vertex_t> & y, double & beta)const{
+  // merge clusters that collapsed or never separated, 
+  // only merge if the estimated critical temperature of the merged vertex is below the current temperature
+  // return true if vertices were merged, false otherwise
+  if(y.size()<2)  return false;
+
+  for(vector<vertex_t>::iterator k=y.begin(); (k+1)!=y.end(); k++){
+    if ( std::abs((k+1)->z - k->z) < 2.e-3 && 
+         std::abs((k+1)->t - k->t) < 2.e-3    ) { 
```

Comment:
is this 2e-3 in some logical relationship with 1e-3 in the other version of ::merge?
better use constants.